### DISCUSSION TITLE
Welcome to futu_algo Discussions!

### DISCUSSION THREAD

**Discussion by billpwchan** [billpwchan/futu_algo#7] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by cantona** [comment]
Finally install all libraries, but still got error:

$ python3 main.py 
/home/test/.local/lib/python3.8/site-packages/pyfolio/pos.py:26: UserWarning: Module "zipline.assets" not found; mutltipliers will not be applied to position notionals.
  warnings.warn(
Traceback (most recent call last):
  File "main.py", line 169, in <module>
    main()
  File "main.py", line 135, in main
    filter_list = [file_name.split("\\")[1][:-3] for file_name in glob.glob(f"./filters/*.py") if
  File "main.py", line 135, in <listcomp>
    filter_list = [file_name.split("\\")[1][:-3] for file_name in glob.glob(f"./filters/*.py") if
IndexError: list index out of range

**Reply by billpwchan** [reply]
Heyyy thanks for the question!

From my understanding, this line essentially is to try get all the available stock filters files in the ./filters/ folder. For instance, in the current code repository, we have MA_simple.py, Price_Threshold.py, etc. In my private repository, I also defined some own filters such as DZX_1B.py, which is not pushed to this code repository. This may cause some issues because, in line 75-84, you can still see those privately defined filters stored in the dictionary, it is just not yet pushed to the github repository.

In short, use the command `py main.py -f Volume_Threshold Price_Threshold` can quickly help you find all stocks that have a monthly average volumn > 10M and monthly average price > 1HKD, and you can also tune these parameters directly in the code (Sorry I wasn't having time to provide a more user-friendly interface! Will do after the exam haha)

Please let me know if this can help you resolve the issue.

**Reply by billpwchan** [reply]
The README wasn't fully updated with latest features that I wrote for this code. 

Just a brief summary of what this code can do now:
1. Connect to both paper trade / real-time trade account in Futu
2. Use pre-defined strategies to HFT using 1M real-time data. Whenever the data is pushed to the client, it will trigger the callback function to apply the strategy you want to use. One example would be `python main.py -s MACD_Cross`, whenever a MACD gold cross appeared, buy 1 lot of shares. 
**However, this needs you to firstly download some stock data and store them in the database! ** To do that, you need to firstly create a folder ./database and use the ./util/database_ddl.sql to create that database. Followed by that, change line 53 in main.py to a customized stock list that you want to focus on trading (Sorry will provide a better way for users to input their selected stocks), such as `['HK.00001', 'HK.00002']`. Lastly, use command `python main.py -u` will help you update 1M and 1D stock data and stored them in both the database and ./data folder. _Note these steps are only needed if you want to use the algo trade feature_
3. Filter stocks based on your selections and send an email defined in the SubscriptionList field in config.ini. Checkout the latest README for how to to setup this file.

**Reply by cantona** [reply]
I am able to run some strategies in futu (ie. EMA_Ribbon, KDJ_Cross, RSI_Threshold)
Previous problems are:
- some strategies and filters file that import are missed in repo 
- on linux there are no win10toast
- linux path compatibility 

patch here:
[fix.txt](https://github.com/billpwchan/futu_algo/files/6161445/fix.txt)

Thanks for sharing this to github its fun. Let's see how we can enhance this to be more useful.

**Reply by cantona** [reply]
sorry, i have questions about your mentioned above:

1) main.py -f Volume_Threshold Price_Threshold  i am able to get a list of stock in output and log/futu_algo_2021-03-18.log., but where are those stock list store at?  how to make use of them?
2) i have already run ":main.py -u update" to have included  (ie HK.00001) data into database, however " main.py -s MACD_Cross" will trade other stock but not only HK.00001, is this normal?
3) i am not understand how to config filter list/custom stock list, right now i am modifying below code to have a filtered stock list 
4) as you mentioned i have to download database if i want to use the algo trade feature, can you tell me which related code are using such database for trading? Because i cannot trace that code.

    if args.strategy:
        # Initialize Strategies
        stock_list = filtered_stock_list if args.filter else data_engine.DatabaseInterface(
            database_path='./database/stock_data.sqlite').get_stock_list()
        stock_list.extend(data_engine.YahooFinanceInterface.get_top_30_hsi_constituents())
        init_day_trading(futu_trade, stock_list, args.strategy)
        futu_trade.display_quota()

4) "main.py -f Volume_Threshold Price_Threshold -s KDJ_Cross" will still include the stock from top 30hsi right? According to the source code

**Reply by billpwchan** [reply]
Let's firstly talk about the fix you provided! That's so useful for making this code more accessible for different users! 

Yet I do need some suggestions on this fix before integrating them:
1. Win10Toast is a failed experiment I did long time ago. Please just remove that entirely hahah
2. My original design is that each user will write their own strategies, that they do not want to share with the public? Do you have any recommendations for separating strategies that are private and public, because I **have to initialize all of them**  (in functions __init_filter and __init_strategy in main.py) based on the input from the argparser, and I cannot dynamically evaluate that string input as an instance on run-time. 
2. For the path fix in different system, would you mind helping me write a simple logic to detect which OS the user is running? and use the path correspondingly. 
3. And definitely, the __init_backtesting feature isn't working at now (You will see the backtesting results are ridiculous hahahah). Will do more debugging before including this feature into the README

**Reply by billpwchan** [reply]
Please see my responses as belows:
> 
> 1. main.py -f Volume_Threshold Price_Threshold  i am able to get a list of stock in output and log/futu_algo_2021-03-18.log., but where are those stock list store at?  how to make use of them?

My design is to send an Email to emails that are defined in config.ini -> SubscriptionList. But before that, you need to setup your own email credentials so that you can send that email to yourself. My setup on my side is 

`Port = 587
Smtp_server = smtp-mail.outlook.com
Sender = billpwchan@hotmail.com
Login = billpwchan@hotmail.com
Password = xxxxxxxxxxxxxxxxx
SubscriptionList = ["billpwchan@hotmail.com"]`

Just to make sure, this password is only used by the smtplib locally, so nobody will see that, and it is safe for you to include them in your local config.ini.

> 2. i have already run ":main.py -u update" to have included  (ie HK.00001) data into database, however " main.py -s MACD_Cross" will trade other stock but not only HK.00001, is this normal?

Please checkout the latest code. Right now the logic will be use the config.ini -> StockList field to define your interested stocks! I've also included a config_template.ini to help you quickly setup your settings lol. But you are right, if you only have ["HK.00001"] in your StockList attribute, you should only be able to trade this single stock using -s option.

> 3. i am not understand how to config filter list/custom stock list, right now i am modifying below code to have a filtered stock list
> 4. as you mentioned i have to download database if i want to use the algo trade feature, can you tell me which related code are using such database for trading? Because i cannot trace that code.
>    if args.strategy:
>    Initialize Strategies
>    stock_list = filtered_stock_list if args.filter else data_engine.DatabaseInterface(
>    database_path='./database/stock_data.sqlite').get_stock_list()
>    stock_list.extend(data_engine.YahooFinanceInterface.get_top_30_hsi_constituents())
>    init_day_trading(futu_trade, stock_list, args.strategy)
>    futu_trade.display_quota()

Please check my latest update. Right now the logic will be 
1. If config.ini -> StockList exists, the program will use this list as the stock_list, and trade them / update data. 
2. If config.ini -> Stocklist = [], then the program will use all available stock codes in your database to trade them / update them. 

Other then that, the database is **not used at all because I haven't think of a use case for that loll**. It is not used for trading, filtering updating or so. But please bear in mind, when you use the `python main.py -u`, whenever a file is downloaded to your ./data/HK.xxxxx, this record will also be inserted to your database. This is mainly defined in ./engines/data_engine.py -> DatabaseInterface. 

> 5. "main.py -f Volume_Threshold Price_Threshold -s KDJ_Cross" will still include the stock from top 30hsi right? According to the source code

Yeap that's correct. Sometimes I'm using those and sometimes I don't, so the version of main.py that I pushed to the code repository might accidently 'comment out' this line because in some cases, I don't want to trade those top 30 hsi. 

Finally, please let me know if you have some suggestions on implementing this software in a more user-friendly way! I often comment out / change lines in the main.py so that to achieve the things that I want to do, but that shouldn't be done this way. Instead, these parameters should be properly defined in the argument / environment variable so that the main.py should only serve as an interface for executing user orders. Just make a pull requests or attach a quick fix in the chat would be very helpful! I will delete the win10toast part completely for now.

**Reply by cantona** [reply]
something like this to compatible to *nix and windows

from pathlib import Path
import glob
[Path(file).name[:-3] for file in glob.glob("./strategies/*.py") if "__init__" not in file and "Strategies" not in file]
[Path(file).name[:-3] for file in glob.glob("./filters/*.py") if "__init__" not in file and "Filters" not in file]

**removed 'f' as no formatted string here

[0001-Fixes-to-deal-with-file-paths-on-Windows-Mac-and-Lin.txt](https://github.com/billpwchan/futu_algo/files/6162797/0001-Fixes-to-deal-with-file-paths-on-Windows-Mac-and-Lin.txt)

**Reply by billpwchan** [reply]
Superb! Updated these in commit d5546c472cba398e4a024761e826e2b951182d24

**Reply by cantona** [reply]
Btw, I tried to use MACD_Cross  for paper trade today, it seems that is not very practical.  Which strategies do you think work the best? If it is possible to share the private strategies? I am very eager to test it.  
Do you think some library ie PyAlgoTrade or backtrader will help to improve futu_algo?
Also, does this auto trade work for US market?

**Reply by billpwchan** [reply]
Yeap definitely. I totally agreed with you, because the MACD is a really simple and basic strategy that solely relies on one simple indicator, and is not effective for high-frequency trading for sure. For the private strategy that I implemented, I haven't backtested them so that I'm really not sure if those are useful strategies or not. Let me share with you once I have fixed my backtesting engine and prove their ROI.

I wasn't aware of libraries you mentioned like PyAlgoTrade and backtrader, but definitely they are useful! For backtesting, I mostly use amibroker instead of actually implementing a backtesting engine in Python from scratch. That's why I am not that engaged in fixing the backtesting bug in this program haha

For US market, sure it works. Just need some tiny modifications following the tutorial on Futu OpenAPI. I believe the key point is to select the correct broker account in your Futu account. in the current code, it directly select the default option, which is the HK market.

**Reply by crosstmh** [reply]
I dont think people will share winning stategies. It's like ask strangers for tons of money. But we can discuss our private trading logic~~ That might be illuminating. For me , I sell cash secured puts and covered calls with company I have faith in ，like Tesla.

**Comment by sreleifs** [comment]
Hi, I am a newbie to python. Thanks this is a great project. However I come across some problems which I tried for some time. I could load the environment in conda.  I thought I could get the stock list from internet (hkex) and fill it by itself. so I remove "StockList = ["HK.00001", "HK.00002"]" .

Q1. I've got the following errors. I would like to ask ,can I get the stocklist from hkex automatically and update its historical data?

Traceback (most recent call last):
  File "C:\Users\nevil\anaconda3\envs\futu_Trade\lib\configparser.py", line 789, in get
    value = d[option]
  File "C:\Users\nevil\anaconda3\envs\futu_Trade\lib\collections\__init__.py", line 898, in __getitem__
    return self.__missing__(key)            # support subclasses that define __missing__
  File "C:\Users\nevil\anaconda3\envs\futu_Trade\lib\collections\__init__.py", line 890, in __missing__
    raise KeyError(key)
KeyError: 'stocklist'

Q2. Even if I add back, I've got another error:

Traceback (most recent call last):
  File "main_backend.py", line 223, in <module>
    main()
  File "main_backend.py", line 202, in main
    stock_list.remove('Stock_Pool')
ValueError: list.remove(x): x not in list

Many thanks. I couldn't get it run other than a GUI prompt.

**Reply by billpwchan** [reply]
Hi,

Thanks for the question. Indeed, it's a bug that I didn't discover previously, and let me tell you a quick walkaround in here. 

So firstly, you still need to put some stock tickers in the `config.ini` file, which I believe you should've already created based on the `config_template.ini` I've provided. You shouldn't be removing this "StockList" parameter in the Config because the program uses this to decide what stocks to process. The rationale here is that FutuAPI limits the number of stock codes that you can download historical k-line, and it's better to keep it within 300 if you are in the first tier. 

Now you should have something like:
```
[TradePreference]
LotSizeMultiplier = 2
MaxPercPerAsset = 10
StockList = ["HK.00001", "HK.00002", ....... (Any stock tickers you like) ]
```

Moving on, just simply remove the line "stock_list.remove("Stock_Pool")" because you don't need it in your first run. 

For your second run once all those historical data has been created, just put this line back in because there will be a folder named `Stock_Pool` in the `data` folder created by using the "Update" command. 

Let me know if you have any question, glad to help~

**Reply by billpwchan** [reply]
Updated the code in https://github.com/billpwchan/futu_algo/commit/3e9b287485560669102491466f2c98d26f3c72ae. Now you should be able to run without any issue. 

Quick tips, you can use the Top HSI Constituents for now
```
['HK.00002', 'HK.02319', 'HK.00267', 'HK.00386', 'HK.00101', 'HK.01038', 'HK.02331', 'HK.00027', 'HK.01398', 'HK.00003', 'HK.01044', 'HK.00992', 'HK.00012', 'HK.09633', 'HK.00883', 'HK.01109', 'HK.01093', 'HK.03690', 'HK.02020', 'HK.00017', 'HK.01810', 'HK.02628', 'HK.02688', 'HK.09618', 'HK.00669', 'HK.02269', 'HK.02018', 'HK.00241', 'HK.09988', 'HK.06098']
```

**Reply by sreleifs** [reply]
thanks for your reply. I haven't expected there's still reply. Many thanks. Keep up your good work!

I tried removed that line first. and I've got this error


FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\nevil\\Documents\\python projects\\futu_algo-master\\data\\Stock_Pool\\ListOfSecurities.csv'

**Reply by sreleifs** [reply]
python main_backend.py --force_update
←[0;30m2022-07-17 03:09:25,054 | 18424 | [open_context_base.py] _send_init_connect_sync:311: InitConnect ok: conn_id=1, host=127.0.0.1, port=11111, user_id=xxxxxxx←[0m
←[0;30m2022-07-17 03:09:25,068 | 18424 | [open_context_base.py] _send_init_connect_sync:311: InitConnect ok: conn_id=2, host=127.0.0.1, port=11111, user_id=xxxxxxx←[0m
Traceback (most recent call last):
  File "main_backend.py", line 223, in <module>
    main()
  File "main_backend.py", line 182, in main
    stock_list = json.loads(config.get('TradePreference', 'StockList'))
  File "C:\Users\nevil\anaconda3\envs\futu_trade\lib\json\__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "C:\Users\nevil\anaconda3\envs\futu_trade\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Users\nevil\anaconda3\envs\futu_trade\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 2 (char 1)

**Reply by billpwchan** [reply]
For the `stock_list = json.loads(config.get('TradePreference', 'StockList'))` issue, it is because you need to have the following content in the config.ini. So once you have the following things in the file, you'll be fine
```
[TradePreference]
LotSizeMultiplier = 2
MaxPercPerAsset = 10
StockList = ['HK.00002', 'HK.02319', 'HK.00267', 'HK.00386', 'HK.00101', 'HK.01038', 'HK.02331', 'HK.00027', 'HK.01398', 'HK.00003', 'HK.01044', 'HK.00992', 'HK.00012', 'HK.09633', 'HK.00883', 'HK.01109', 'HK.01093', 'HK.03690', 'HK.02020', 'HK.00017', 'HK.01810', 'HK.02628', 'HK.02688', 'HK.09618', 'HK.00669', 'HK.02269', 'HK.02018', 'HK.00241', 'HK.09988', 'HK.06098']
```

For the FileNotFoundError, may I know which line causes the error? As far as I remembered, this error should not be thrown because the file "ListOfSecurities.xlsx" and "ListOfSecurities.csv" will be created after downloading the data from HKEX. 

Just wanna double check, did you have the the following folder structure similar like this?
data
|
-----HK.00001
-----HK.00002
.....
-----Stock Pools 
       \--------ListOfSecurities.xlsx
       \--------ListOfSecurities.csv

**Reply by sreleifs** [reply]
Hi,

I tried to move to ubuntu to try. But I got stuck connecting to futu. I need to make sure I can connect to Futu before connecting to your apps. Do you mind if you can give me some ideas? Million thanks.

I ./FutuOpenD -> listening port 11111
2. run my python3 test.py to try

2022-07-24 12:37:06,939 | 37871 | [network_manager.py] _on_packet:647: Recv: conn_id=1; proto_id=1001; serial_no=6215; data_len=0; msg=PacketErr.Timeout;
2022-07-24 12:37:06,940 | 37871 | [open_context_base.py] _send_init_connect_sync:316: InitConnect fail: PacketErr.Timeout

**Reply by billpwchan** [reply]
Hi,

Though I didn't test Futu OpenD on Ubuntu previously, I think as long as you follow the steps listed on the FutuOpenD website you shall be able to log in and run the service w/o issue. 

**Website:** https://www.futunn.com/hk/download/OpenAPI

![image](https://user-images.githubusercontent.com/25129943/180654540-bcab8fbf-2ab4-4d14-9cb2-56876af2c2af.png)
